### DISCUSSION TITLE
How to subscribe events

### DISCUSSION THREAD

**Discussion by chabashilah** [polkascan/py-substrate-interface#306] [Q&A] [answered]
Hi,

I'm looking for a way to subscribe a runtime events using py-substrate-interface but couldn't any document or example regarding it.
How can I do that?

**Comment by arjanz** [comment]
This is not possible at the moment, as the current library is only intended be used in a synchronous context. Eventually we would also like to have a async version, for async Python apps.

**Comment by arjanz** [comment]
The only subscription is currently supports is to wait for the result when sending an extrinsic (the `wait_for_inclusion` and `wait_for_finalization` kwargs), but this will block execution until confirmation is given.

**Comment by chabashilah** [comment]
Thank you for your reply! I understood that.

**Comment by arjanz** [comment, marked as answer]
Good news, since latest release https://github.com/polkascan/py-substrate-interface/releases/tag/v0.12.1 storage subscriptions are possible, so for example to subscribe to events:

```python
def subscription_handler(obj, update_nr, subscription_id):

    # Do something with the update
    print('Events:', obj.value)

    # The execution will block until an arbitrary value is returned, which will be the result of the `query`
    if update_nr > 5:
        return {'updates': update_nr, 'message': 'processed events'}


result = substrate.query("System", "Events", [], subscription_handler=subscription_handler)
```
